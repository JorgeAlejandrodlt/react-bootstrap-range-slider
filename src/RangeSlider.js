/**
 * MIT License
 *
 * Copyright (c) 2020 Jason Wilson
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { useBootstrapPrefix } from 'react-bootstrap/ThemeProvider';
import { Tooltip } from 'react-bootstrap';
import classNames from 'classnames';

const CLASS_PREFIX = 'react-bootstrap-range-slider';

let idIndex = 0;

const RangeSlider = React.forwardRef(({
  bsPrefix,
  size,
  className,
  id,
  disabled = false,
  value,
  onChange,
  min = 0,
  max = 100,
  step,
  tooltip,
  tooltipPlacement = 'bottom',
  tooltipDisplay = 'auto',
  tooltipLabel,
  tooltipId,
  tooltipProps,
  inputProps = {},
}, ref) => {

  const prefix = useBootstrapPrefix(bsPrefix, CLASS_PREFIX);

  const classes = classNames(
    className,
    prefix,
    size && `${prefix}--${size}`,
    disabled && 'disabled',
  );

  const inputEl = (
    <input
      type="range"
      className={classes}
      id={id}
      value={value}
      min={min}
      max={max}
      step={step}
      onChange={onChange}
      disabled={disabled}
      ref={ref}
      style={{
        '--value': value,
      }}
      {...inputProps}
    />
  );

  if (tooltip) {

    const wrapClasses = classNames(
      `${prefix}__wrap`,
      size && `${prefix}__wrap-${size}`,
    );

    const tooltipClasses = classNames(
      `${prefix}__tooltip`,
      tooltipPlacement && `${prefix}__tooltip--${tooltipPlacement}`,
      tooltipDisplay && `${prefix}__tooltip--${tooltipDisplay}`,
    );

    const ttId = tooltipId || (id ? `${id}__tooltip` : `${CLASS_PREFIX}__tooltip--${++idIndex}`);
 
    const fract = value / (max - min);
    const percentLeft = fract * 100;
    const fractFromCentre = (fract - 0.5) * 2;
    const adjustment = fractFromCentre * -8; // Half thumb width

    return (
      <span className={wrapClasses}>

        {inputEl}

        <div
          className={tooltipClasses}
          style={{
            left: `calc(${percentLeft}% + ${adjustment}px)`,
          }}
        >

          <Tooltip
            id={ttId}
            placement={tooltipPlacement}
            className='show'
            {...tooltipProps}
          >

            {tooltipLabel ? tooltipLabel(value) : value}

          </Tooltip>

        </div>

      </span>
    );

  } else {

    return inputEl;

  }

});

RangeSlider.propTypes = {
  bsPrefix: PropTypes.string,
  size: PropTypes.oneOf(['sm', 'lg']),
  className: PropTypes.string,
  id: PropTypes.string,
  disabled: PropTypes.bool,
  value: PropTypes.number.isRequired,
  min: PropTypes.number,
  max: PropTypes.number,
  step: PropTypes.number,
  onChange: PropTypes.func.isRequired,
  tooltip: PropTypes.bool,
  tooltipPlacement: PropTypes.oneOf(['top', 'bottom']),
  tooltipDisplay: PropTypes.oneOf(['auto', 'on']),
  tooltipLabel: PropTypes.func,
};

export default RangeSlider;
