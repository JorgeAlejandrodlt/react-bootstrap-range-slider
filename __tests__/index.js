/**
 * MIT License
 *
 * Copyright (c) 2020 Jason Wilson
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import renderer from 'react-test-renderer'

import RangeSlider from '../src/index';

describe('RangeSlider', () => {

  it('should render', () => {

    const tree = renderer
      .create(
        <RangeSlider
          value={25}
          onChange={() => {}}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

  it('should render with tooltip', () => {

    const tree = renderer
      .create(
        <RangeSlider
          value={25}
          onChange={() => {}}
          tooltip
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

  it('should render with tooltip label', () => {

    const tree = renderer
      .create(
        <RangeSlider
          value={25}
          onChange={() => {}}
          tooltip
          tooltipLabel={value => `${value} km`}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

  it('should render with tooltip placement = top', () => {

    const tree = renderer
      .create(
        <RangeSlider
          value={25}
          onChange={() => {}}
          tooltip
          tooltipPlacement='top'
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

  it('should render with tooltipId', () => {

    const tree = renderer
      .create(
        <RangeSlider
          value={25}
          onChange={() => {}}
          tooltip
          tooltipId='a-tooltip-id'
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

  it('should render with id', () => {

    const tree = renderer
      .create(
        <RangeSlider
          value={25}
          onChange={() => {}}
          id='a-slider-id'
          tooltip
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

  it('should render with size = sm', () => {

    const tree = renderer
      .create(
        <RangeSlider
          value={25}
          onChange={() => {}}
          size='sm'
          tooltip
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

  it('should render with bsPrefix', () => {

    const tree = renderer
      .create(
        <RangeSlider
          value={25}
          onChange={() => {}}
          bsPrefix='a-prefix'
          size='sm'  
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

  it('should render when disabled', () => {

    const tree = renderer
      .create(
        <RangeSlider
          value={25}
          onChange={() => {}}
          disabled
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

  });

});
