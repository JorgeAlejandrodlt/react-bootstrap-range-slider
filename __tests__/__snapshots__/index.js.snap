// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider should render 1`] = `
<input
  className="react-bootstrap-range-slider"
  disabled={false}
  max={100}
  min={0}
  onChange={[Function]}
  style={
    Object {
      "--value": 25,
    }
  }
  type="range"
  value={25}
/>
`;

exports[`RangeSlider should render when disabled 1`] = `
<input
  className="react-bootstrap-range-slider disabled"
  disabled={true}
  max={100}
  min={0}
  onChange={[Function]}
  style={
    Object {
      "--value": 25,
    }
  }
  type="range"
  value={25}
/>
`;

exports[`RangeSlider should render with bsPrefix 1`] = `
<input
  className="a-prefix a-prefix--sm"
  disabled={false}
  max={100}
  min={0}
  onChange={[Function]}
  style={
    Object {
      "--value": 25,
    }
  }
  type="range"
  value={25}
/>
`;

exports[`RangeSlider should render with id 1`] = `
<span
  className="react-bootstrap-range-slider__wrap"
>
  <input
    className="react-bootstrap-range-slider"
    disabled={false}
    id="a-slider-id"
    max={100}
    min={0}
    onChange={[Function]}
    style={
      Object {
        "--value": 25,
      }
    }
    type="range"
    value={25}
  />
  <div
    className="react-bootstrap-range-slider__tooltip react-bootstrap-range-slider__tooltip--bottom react-bootstrap-range-slider__tooltip--auto"
    style={
      Object {
        "left": "calc(25% + 4px)",
      }
    }
  >
    <div
      className="show tooltip bs-tooltip-bottom"
      id="a-slider-id__tooltip"
      role="tooltip"
      x-placement="bottom"
    >
      <div
        className="arrow"
      />
      <div
        className="tooltip-inner"
      >
        25
      </div>
    </div>
  </div>
</span>
`;

exports[`RangeSlider should render with size = sm 1`] = `
<span
  className="react-bootstrap-range-slider__wrap react-bootstrap-range-slider__wrap-sm"
>
  <input
    className="react-bootstrap-range-slider react-bootstrap-range-slider--sm"
    disabled={false}
    max={100}
    min={0}
    onChange={[Function]}
    style={
      Object {
        "--value": 25,
      }
    }
    type="range"
    value={25}
  />
  <div
    className="react-bootstrap-range-slider__tooltip react-bootstrap-range-slider__tooltip--bottom react-bootstrap-range-slider__tooltip--auto"
    style={
      Object {
        "left": "calc(25% + 4px)",
      }
    }
  >
    <div
      className="show tooltip bs-tooltip-bottom"
      id="react-bootstrap-range-slider__tooltip--4"
      role="tooltip"
      x-placement="bottom"
    >
      <div
        className="arrow"
      />
      <div
        className="tooltip-inner"
      >
        25
      </div>
    </div>
  </div>
</span>
`;

exports[`RangeSlider should render with tooltip 1`] = `
<span
  className="react-bootstrap-range-slider__wrap"
>
  <input
    className="react-bootstrap-range-slider"
    disabled={false}
    max={100}
    min={0}
    onChange={[Function]}
    style={
      Object {
        "--value": 25,
      }
    }
    type="range"
    value={25}
  />
  <div
    className="react-bootstrap-range-slider__tooltip react-bootstrap-range-slider__tooltip--bottom react-bootstrap-range-slider__tooltip--auto"
    style={
      Object {
        "left": "calc(25% + 4px)",
      }
    }
  >
    <div
      className="show tooltip bs-tooltip-bottom"
      id="react-bootstrap-range-slider__tooltip--1"
      role="tooltip"
      x-placement="bottom"
    >
      <div
        className="arrow"
      />
      <div
        className="tooltip-inner"
      >
        25
      </div>
    </div>
  </div>
</span>
`;

exports[`RangeSlider should render with tooltip label 1`] = `
<span
  className="react-bootstrap-range-slider__wrap"
>
  <input
    className="react-bootstrap-range-slider"
    disabled={false}
    max={100}
    min={0}
    onChange={[Function]}
    style={
      Object {
        "--value": 25,
      }
    }
    type="range"
    value={25}
  />
  <div
    className="react-bootstrap-range-slider__tooltip react-bootstrap-range-slider__tooltip--bottom react-bootstrap-range-slider__tooltip--auto"
    style={
      Object {
        "left": "calc(25% + 4px)",
      }
    }
  >
    <div
      className="show tooltip bs-tooltip-bottom"
      id="react-bootstrap-range-slider__tooltip--2"
      role="tooltip"
      x-placement="bottom"
    >
      <div
        className="arrow"
      />
      <div
        className="tooltip-inner"
      >
        25 km
      </div>
    </div>
  </div>
</span>
`;

exports[`RangeSlider should render with tooltip placement = top 1`] = `
<span
  className="react-bootstrap-range-slider__wrap"
>
  <input
    className="react-bootstrap-range-slider"
    disabled={false}
    max={100}
    min={0}
    onChange={[Function]}
    style={
      Object {
        "--value": 25,
      }
    }
    type="range"
    value={25}
  />
  <div
    className="react-bootstrap-range-slider__tooltip react-bootstrap-range-slider__tooltip--top react-bootstrap-range-slider__tooltip--auto"
    style={
      Object {
        "left": "calc(25% + 4px)",
      }
    }
  >
    <div
      className="show tooltip bs-tooltip-top"
      id="react-bootstrap-range-slider__tooltip--3"
      role="tooltip"
      x-placement="top"
    >
      <div
        className="arrow"
      />
      <div
        className="tooltip-inner"
      >
        25
      </div>
    </div>
  </div>
</span>
`;

exports[`RangeSlider should render with tooltipId 1`] = `
<span
  className="react-bootstrap-range-slider__wrap"
>
  <input
    className="react-bootstrap-range-slider"
    disabled={false}
    max={100}
    min={0}
    onChange={[Function]}
    style={
      Object {
        "--value": 25,
      }
    }
    type="range"
    value={25}
  />
  <div
    className="react-bootstrap-range-slider__tooltip react-bootstrap-range-slider__tooltip--bottom react-bootstrap-range-slider__tooltip--auto"
    style={
      Object {
        "left": "calc(25% + 4px)",
      }
    }
  >
    <div
      className="show tooltip bs-tooltip-bottom"
      id="a-tooltip-id"
      role="tooltip"
      x-placement="bottom"
    >
      <div
        className="arrow"
      />
      <div
        className="tooltip-inner"
      >
        25
      </div>
    </div>
  </div>
</span>
`;
